[package] Make it possible to re-save a PackageImporter module

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65101

As title. Previously this was guarded against for implementation
simplicity, as we didn't really think there was a use case for saving a
mangled module name directly.

But people started doing stuff like:
```
exporter.save_module(my_imported_obj.__module__)
```
which implicitly passes along the mangled module name.

This PR makes it so that given `PackageImporter` instance can always
import modules that it created, and changes `PackageExporter` to
properly demangle the resulting module name when writing the package to
the export archive.

Differential Revision: [D30975712](https://our.internmc.facebook.com/intern/diff/D30975712)

suo committed:
[package] Make it possible to re-save a PackageImporter module

As title. Previously this was guarded against for implementation
simplicity, as we didn't really think there was a use case for saving a
mangled module name directly.

But people started doing stuff like:
```
exporter.save_module(my_imported_obj.__module__)
```
which implicitly passes along the mangled module name.

This PR makes it so that given `PackageImporter` instance can always
import modules that it created, and changes `PackageExporter` to
properly demangle the resulting module name when writing the package to
the export archive.

[ghstack-poisoned]

suo committed:
Update on "[package] Make it possible to re-save a PackageImporter module"


As title. Previously this was guarded against for implementation
simplicity, as we didn't really think there was a use case for saving a
mangled module name directly.

But people started doing stuff like:
```
exporter.save_module(my_imported_obj.__module__)
```
which implicitly passes along the mangled module name.

This PR makes it so that given `PackageImporter` instance can always
import modules that it created, and changes `PackageExporter` to
properly demangle the resulting module name when writing the package to
the export archive.

[ghstack-poisoned]

SplitInfinity commented on test/package/test_mangling.py:
```diff
@@ -96,6 +96,15 @@ def test_unique_module_names(self):
+        a_mangled = a.mangle("foo.bar")
+        # Since `a` mangled this string, it should demangle properly.
+        self.assertEqual(a.demangle(a_mangled), "foo.bar")
+        # Since `b` did not mangle this string, demangling should leave it along.
```
```suggestion
        # Since `b` did not mangle this string, demangling should leave it alone.
```

SplitInfinity commented on test/package/test_save_load.py:
```diff
@@ -235,6 +235,30 @@ def make_exporter():
+
+        # Should export without error.
+        buffer2 = BytesIO()
+        with PackageExporter(buffer2, importer=(importer, sys_importer)) as exporter:
+            exporter.intern("**")
+            exporter.save_module(imported_obj2_module)
```
If this works without errors, how do you know that `importer` was used to import `imported_obj2_module` and not `sys_importer`? That's what this test is trying to test, right?

suo replied:
yeah, make sense to actually check it lol

SplitInfinity commented on torch/package/package_exporter.py:
```diff
@@ -511,6 +514,7 @@ def _intern_module(
                 is_package=is_package,
                 error=packaging_error,
                 error_context=error_context,
+                provided=True,
```
I'm guessing this is necessary to prevent `add_dependency` from adding another node for this module in the dependency graph, but it is counterintuitive for this to be "provided" since there is a packaging error attached. The graph will fail validation and won't be executed.

suo replied:
"provided" just means "don't process it again" (probably should have a better name). Previously this was added in `save_module` above but I deleted that line so I had to add it back here.

SplitInfinity reviewed (commented):
🥇

Psoffie009 commented:
> 🥇

suo commented:
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30975712).

Psoffie009 commented:
> CI Flow Status

Psoffie009 commented:
> Stack from [ghstack](https://github.com/ezyang/ghstack):
> 
> * **->** [[package] Make it possible to re-save a PackageImporter module #65101](https://github.com/pytorch/pytorch/pull/65101)
> 
> As title. Previously this was guarded against for implementation
> simplicity, as we didn't really think there was a use case for saving a
> mangled module name directly.
> 
> But people started doing stuff like:
> 
> ```
> exporter.save_module(my_imported_obj.__module__)
> ```
> 
> which implicitly passes along the mangled module name.
> 
> This PR makes it so that given `PackageImporter` instance can always
> import modules that it created, and changes `PackageExporter` to
> properly demangle the resulting module name when writing the package to
> the export archive.
> 
> Differential Revision: [D30975712](https://our.internmc.facebook.com/intern/diff/D30975712)

suo committed:
Update on "[package] Make it possible to re-save a PackageImporter module"


As title. Previously this was guarded against for implementation
simplicity, as we didn't really think there was a use case for saving a
mangled module name directly.

But people started doing stuff like:
```
exporter.save_module(my_imported_obj.__module__)
```
which implicitly passes along the mangled module name.

This PR makes it so that given `PackageImporter` instance can always
import modules that it created, and changes `PackageExporter` to
properly demangle the resulting module name when writing the package to
the export archive.

Differential Revision: [D30975712](https://our.internmc.facebook.com/intern/diff/D30975712)

[ghstack-poisoned]

Psoffie009 commented:
> ## 🔗 Helpful links
> * [🧪](https://hud.pytorch.org/pr/65101)  **See artifacts and rendered test results [at hud.pytorch.org/pr/65101](https://hud.pytorch.org/pr/65101)**
> * [↩️  [fb-only] Re-run with SSH instructions](https://github.com/pytorch/pytorch/wiki/Debugging-using-with-ssh-for-Github-Actions)
> * [🔧  Opt-in to CIFlow to control what jobs run on your PRs](https://github.com/pytorch/pytorch/wiki/Continuous-Integration#user-guide)
> 
> ## 💊 CI failures summary and remediations
> As of commit [71e633a](https://github.com/pytorch/pytorch/commit/71e633af61fdd726cf42432a970c05565fe11da2) (more details [on the Dr. CI page](https://our.intern.facebook.com/intern/opensource/dr_ci/commit/167825833786582/71e633af61fdd726cf42432a970c05565fe11da2/)):
> 
> * **2/2** failures introduced in this PR
> 
> ### 2 failures _not_ recognized by patterns:
> Job	Step	Action
> ![GitHub Actions](https://avatars.githubusercontent.com/in/15368?s=12) linux-xenial-cuda11.3-py3.6-gcc7 / build	Unknown	[🔁 rerun](https://our.intern.facebook.com/intern/opensource/ci/restart/4317076601679876/)
> ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) pytorch_linux_xenial_py3_6_gcc5_4_build	Build	[🔁 rerun](https://our.intern.facebook.com/intern/opensource/ci/restart/975095040018577/)
> This comment was automatically generated by [Dr. CI](https://code.facebook.com/ci/dr-ci-info/) (expand for details).
> Click[ here ](https://our.intern.facebook.com/intern/opensource/ci/regenerate_comment/941206170158312/) to manually regenerate this comment.

suo commented:
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30975712).

suo committed:
Update on "[package] Make it possible to re-save a PackageImporter module"


As title. Previously this was guarded against for implementation
simplicity, as we didn't really think there was a use case for saving a
mangled module name directly.

But people started doing stuff like:
```
exporter.save_module(my_imported_obj.__module__)
```
which implicitly passes along the mangled module name.

This PR makes it so that given `PackageImporter` instance can always
import modules that it created, and changes `PackageExporter` to
properly demangle the resulting module name when writing the package to
the export archive.

Differential Revision: [D30975712](https://our.internmc.facebook.com/intern/diff/D30975712)

[ghstack-poisoned]

suo commented:
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30975712).

suo committed:
Update on "[package] Make it possible to re-save a PackageImporter module"


As title. Previously this was guarded against for implementation
simplicity, as we didn't really think there was a use case for saving a
mangled module name directly.

But people started doing stuff like:
```
exporter.save_module(my_imported_obj.__module__)
```
which implicitly passes along the mangled module name.

This PR makes it so that given `PackageImporter` instance can always
import modules that it created, and changes `PackageExporter` to
properly demangle the resulting module name when writing the package to
the export archive.

Differential Revision: [D30975712](https://our.internmc.facebook.com/intern/diff/D30975712)

[ghstack-poisoned]

suo commented:
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30975712).

suo committed:
Update on "[package] Make it possible to re-save a PackageImporter module"


As title. Previously this was guarded against for implementation
simplicity, as we didn't really think there was a use case for saving a
mangled module name directly.

But people started doing stuff like:
```
exporter.save_module(my_imported_obj.__module__)
```
which implicitly passes along the mangled module name.

This PR makes it so that given `PackageImporter` instance can always
import modules that it created, and changes `PackageExporter` to
properly demangle the resulting module name when writing the package to
the export archive.

Differential Revision: [D30975712](https://our.internmc.facebook.com/intern/diff/D30975712)

[ghstack-poisoned]

suo commented:
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30975712).

suo committed:
Update on "[package] Make it possible to re-save a PackageImporter module"


As title. Previously this was guarded against for implementation
simplicity, as we didn't really think there was a use case for saving a
mangled module name directly.

But people started doing stuff like:
```
exporter.save_module(my_imported_obj.__module__)
```
which implicitly passes along the mangled module name.

This PR makes it so that given `PackageImporter` instance can always
import modules that it created, and changes `PackageExporter` to
properly demangle the resulting module name when writing the package to
the export archive.

Differential Revision: [D30975712](https://our.internmc.facebook.com/intern/diff/D30975712)

[ghstack-poisoned]

suo commented:
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30975712).

suo committed:
Update on "[package] Make it possible to re-save a PackageImporter module"


As title. Previously this was guarded against for implementation
simplicity, as we didn't really think there was a use case for saving a
mangled module name directly.

But people started doing stuff like:
```
exporter.save_module(my_imported_obj.__module__)
```
which implicitly passes along the mangled module name.

This PR makes it so that given `PackageImporter` instance can always
import modules that it created, and changes `PackageExporter` to
properly demangle the resulting module name when writing the package to
the export archive.

Differential Revision: [D30975712](https://our.internmc.facebook.com/intern/diff/D30975712)

[ghstack-poisoned]

suo commented:
@suo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30975712).
